flexmeasures/data/models/planning/tests/test_solver.py

```diff
@@ -1297,3 +1298,205 @@ def test_build_device_soc_values(caplog, soc_values, log_message):
+            "Test battery",
+            False,
+            False,
+            "10 kW",
+            None,
+            # from the flex model field 'production-capacity' (a quantity)
+            [-0.01] * 24 * 4,
+            # from the asset attribute 'capacity_in_mw'
+            [2] * 24 * 4,
+        ),
+        (
+            "Test battery",
+            False,
+            False,
+            "10 kW",
+            "100 kW",
+            # from the flex model field 'production-capacity' (a quantity)
+            [-0.01] * 24 * 4,
+            # from the flex model field 'consumption-capacity' (a quantity)
+            [0.1] * 24 * 4,
+        ),
+    ],
+)
+def test_battery_power_capacity_as_sensor(
+    db,
+    add_battery_assets,
+    add_inflexible_device_forecasts,
+    capacity_sensors,
+    battery_name,
+    production_sensor,
+    consumption_sensor,
+    production_quantity,
+    consumption_quantity,
+    expected_production,
+    expected_consumption,
+):
+
+    epex_da, battery = get_sensors_from_db(
+        add_battery_assets, battery_name=battery_name
+    )
+
+    tz = pytz.timezone("Europe/Amsterdam")
+    start = tz.localize(datetime(2015, 1, 2))
+    end = tz.localize(datetime(2015, 1, 3))
+    resolution = timedelta(minutes=15)
+    soc_at_start = 10
+
+    flex_model = {
+        "soc-at-start": soc_at_start,
+        "roundtrip-efficiency": "100%",
+        "prefer-charging-sooner": False,
+    }
+
+    if production_sensor:
+        flex_model["production-capacity"] = {
+            "sensor": capacity_sensors["production"].id
+        }
+
+    if consumption_sensor:
+        flex_model["consumption-capacity"] = {
+            "sensor": capacity_sensors["consumption"].id
+        }
+
+    if production_quantity is not None:
+        flex_model["production-capacity"] = production_quantity
+
+    if consumption_quantity is not None:
+        flex_model["consumption-capacity"] = consumption_quantity
+
+    scheduler: Scheduler = StorageScheduler(
+        battery, start, end, resolution, flex_model=flex_model
+    )
+
+    data_to_solver = scheduler._prepare()
+    device_constraints = data_to_solver[5][0]
+
+    assert all(device_constraints["derivative min"].values == expected_production)
+    assert all(device_constraints["derivative max"].values == expected_consumption)
+
+
+@pytest.mark.parametrize(
+    "flex_model_field,device_constraint",
+    [
+        ("charging-efficiency", "derivative up efficiency"),
+        ("discharging-efficiency", "derivative down efficiency"),
+    ],
+)
+def test_dis_charging_efficiency_as_sensor(
+    db,
+    add_battery_assets,
+    add_inflexible_device_forecasts,
+    add_market_prices,
+    efficiency_sensors,
+    device_constraint,
+    flex_model_field,
+):
+    """
+    The efficiency sensor defined an efficiency of 90% for 23h of the 24h of the schedule. The last
+    hour should use the roundtrip-efficiency of 92.16% (charge and discharge efficiencies of 96%).
```

Comment:
We should fall back to the charging-efficiency and discharging-efficiency db model attributes instead.

If these also don't exist, we could choose to assume 100%. Alternatively, we could take the last known recorded efficiency (most recent event before the missing event). That would effectively constitute an on-the-fly naive (method=last) forecast.

Reply:
In the last commit, I stopped to rely on the `roundtrip-efficiency` for the fallback value and instead rely on the asset attribute.

Reply:
I don't know which commit you refer to. I still see the roundtrip-efficiency is used as a fallback. Specifically, here:
```
        # if dis/charging efficiency pair is not defined then the roundtrip efficiency is used
        if not (
            "charging_efficiency" in self.flex_model
            or self.sensor.has_attribute("charging-efficiency")
        ):
            charging_efficiency = roundtrip_efficiency**0.5
            discharging_efficiency = roundtrip_efficiency**0.5
```

Reply:
As the `(dis)charging-efficiency` fields cannot be used in combination with the `roundtrip-efficiency`, if the `charging_efficiency` is not defined either in the flex-model or the asset attribute, we get the value from the `roundtrip-efficiency`. If the even the `roundtrip-efficiency` wasn't defined, its value would default to 100%.